### PR TITLE
Improve robustness of journal plugin

### LINKED ISSUE #878: Linux systemd journal plugin type casting is broken
As noted in the [systemd journal field documentation](https://github.com/systemd/systemd/blob/main/man/systemd.journal-fields.xml#L127-L131), systemd does not validate values of structured journal fields. The journal plugin currently breaks if it encounters unexpected values.

### PR DESCRIPTION
**Pull request by JSCU-CNI**
This PR fixes #878. We added some sanity checks to make sure the journal plugin does not break if it encounters values that do not match the format we expected.

### CODE REVIEW

File: dissect/target/plugins/os/unix/log/journal.py [hunk 1]
```diff
@@ -259,16 +262,26 @@
...
 
 
 class JournalFile:
     """Parse Systemd Journal file format.
 
     References:
-        - https://github.com/systemd/systemd/blob/206f0f397edf1144c63a158fb30f496c3e89f256/docs/JOURNAL_FILE_FORMAT.md
-        - https://github.com/libyal/dtformats/blob/c4fc2b8102702c64b58f145971821986bf74e6c0/documentation/Systemd%20journal%20file%20format.asciidoc
-    """  # noqa: E501
+        - https://github.com/systemd/systemd/blob/main/docs/JOURNAL_FILE_FORMAT.md
```
Comment (Zawadidone) [inline dissect/target/plugins/os/unix/log/journal.py:282]: Can this be changed to the specific commit that you read to create this PR? Without this it isn't clear which version/commit of the format was used to create these modifications.

Note I am not a reviewer of this PR or an author of this project.

File: dissect/target/plugins/os/unix/log/journal.py [hunk 2]
```diff
@@ -259,16 +262,26 @@
 
 def get_optional(value: str, to_type: Callable):
```
Comment (Schamper) [inline dissect/target/plugins/os/unix/log/journal.py:263]: Can you add a return type?

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/target/plugins/os/unix/log/journal.py (hunk 1 above)
Comment (Schamper) [reply dissect/target/plugins/os/unix/log/journal.py:284]: Can you also address this? @JSCU-CNI. Using `main` may result in pointing to different/nonexistent documentation in the future.

File: dissect/target/plugins/os/unix/log/journal.py [hunk 3]
```diff
@@ -273,136 +288,138 @@ class JournalFile:
...
-                    entry_array_object = c_journal.EntryArrayObject(self.fh)
-
-                for entry_object_offset in entry_array_object.entry_object_offsets:
-                    # Check if the offset is not zero and points to nothing
-                    if entry_object_offset:
-                        yield entry_object_offset
-
-                offset = entry_array_object.next_entry_array_offset
+        if self.header.signature != c_journal.HEADER_SIGNATURE:
+            raise ValueError(f"Journal file has invalid magic header '{str(self.header.signature)}'")
```
Comment (Schamper) [inline dissect/target/plugins/os/unix/log/journal.py:298]:
```suggestion
            raise ValueError(f"Journal file has invalid magic header: {self.header.signature!r}'")
```

File: dissect/target/plugins/os/unix/log/journal.py [hunk 4]
```diff
@@ -273,136 +288,138 @@ class JournalFile:
...
-        self.header = c_journal.Header(self.fh)
-        self.signature = "".join(chr(c) for c in self.header.signature)
-        self.entry_array_offset = self.header.entry_array_offset
 
-    def entry_object_offsets(self) -> Iterator[int]:
-        """Read object entry arrays."""
+        try:
+            self.header = c_journal.Header(self.fh)
+        except EOFError as e:
+            raise ValueError(f"Invalid Systemd Journal file: {str(e)}")
```
Comment (Schamper) [inline dissect/target/plugins/os/unix/log/journal.py:295]:
```suggestion
            raise ValueError(f"Invalid systemd Journal file: {e}")
```

**Review by Schamper — CHANGES_REQUESTED** [review]
Some small things

File: dissect/target/plugins/os/unix/log/journal.py [hunk 5]
```diff
@@ -273,136 +288,138 @@ class JournalFile:
...
 
     def __iter__(self) -> Iterator[dict[str, int | str]]:
         "Iterate over the entry objects to read payloads."
 
-        for offset in self.entry_object_offsets():
+        offset = self.header.entry_array_offset
+        while offset != 0:
             self.fh.seek(offset)
 
+            if int.from_bytes(self.fh.read(1), "little") != c_journal.ObjectType.OBJECT_ENTRY_ARRAY:
```
Comment (Schamper) [inline dissect/target/plugins/os/unix/log/journal.py:314]:
```suggestion
            if self.fh.read(1)[0] != c_journal.ObjectType.OBJECT_ENTRY_ARRAY:
```

File: dissect/target/plugins/os/unix/log/journal.py [hunk 6]
```diff
@@ -273,136 +288,138 @@ class JournalFile:
...
-            for item in entry.items:
-                try:
-                    self.fh.seek(item.object_offset)
+        event = {"ts": ts.from_unix_us(entry.realtime)}
+        for item in entry.items:
+            try:
+                self.fh.seek(item.object_offset)
 
-                    object = c_journal.ObjectHeader(self.fh)
+                if int.from_bytes(self.fh.read(1), "little") != c_journal.ObjectType.OBJECT_DATA:
```
Comment (Schamper) [inline dissect/target/plugins/os/unix/log/journal.py:347]:
```suggestion
                if self.fh.read(1)[0] != c_journal.ObjectType.OBJECT_DATA:
```

**Review by Schamper — CHANGES_REQUESTED** [review]

**Review by Schamper — APPROVED** [review]

**Pull request merged**